data/src/bouncer.rs

```diff
@@ -0,0 +1,146 @@
+use std::hash::Hash;
+use std::cmp::Ordering;
+
+use std::str::FromStr;
+
+use irc::proto::parse::{Error as ParseError, tagstr};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Eq, PartialEq, Default, Serialize, Deserialize)]
+pub enum NetworkState {
+    Connected,
+    Connecting,
+    #[default]
+    Disconnected,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Unable to parse bouncer network: {0}")]
+    Parse(#[from] ParseError),
+    #[error("Bouncer network missing field: {0}")]
+    MissingField(&'static str),
+    #[error("Invalid network state: {0}")]
+    InvalidState(String),
+}
+
+impl FromStr for NetworkState {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "connected" => Ok(Self::Connected),
+            "connecting" => Ok(Self::Connecting),
+            "disconnected" => Ok(Self::Disconnected),
+            s => Err(Error::InvalidState(s.to_owned())),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+pub struct BouncerNetwork {
+    pub id: String,
+    pub name: String,
+    pub host: Option<String>,
+    pub state: NetworkState,
+    pub port: Option<u16>,
+    pub use_tls: Option<bool>,
+    pub pass: Option<String>,
+    pub nickname: Option<String>,
+    pub realname: Option<String>,
+    pub error: Option<String>,
+}
+
+// for ordering, we try to order lexiographically by name, and then check ID
+// So ID equality must imply name equality
+impl Ord for BouncerNetwork {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // case sensitive first, then insensitive, then ID
+        let case_insensitive_ordering =
+            self.name.to_lowercase().cmp(&other.name.to_lowercase());
+
+        if !matches!(case_insensitive_ordering, Ordering::Equal) {
+            return case_insensitive_ordering;
+        }
+
+        let case_sensitive_ordering = self.name.cmp(&other.name);
+        if !matches!(case_sensitive_ordering, Ordering::Equal) {
+            return case_sensitive_ordering;
+        }
+
+        self.id.cmp(&other.id)
```

Comment:
[then_with](https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then_with) makes composing these more functional and clean